### PR TITLE
use lax.cond operator for gradient accumulation optimizer

### PR DESCRIPTION
**Pull request by shawnwang18**
Use `lax.cond` operator for gradient accumulation optimizer. 

Original while operator will introduce large memory overhead, we can save great memory costs by using `lax.cond`. For an example PAXML config (TP=2, DP=4, Scan=On, 24 layers), this change can reduce peak memory costs from 85GB to about 52GB. 

For the issue described by previous comments: 

```
    # PAX makes use of vectorized map for repeated layers. XLA currently doesn't
    # handle conds with-in vmap well and thus calls into both branches with a
    # select. Here we rewrite a lax.cond as while_loop to get around this issue
    # and get faster step time.
```

Needs to verify whether the issue still exist, and there is, should fix it instead of using while operator to work around.

### CODE REVIEW

**Review by zhangqiaorjc — APPROVED** [review]

**Review by zhangqiaorjc — APPROVED** [review]

File: praxis/optimizers.py
Comment (akuegel) [inline praxis/optimizers.py:2758]: Can you please add   # pytype: disable=attribute-error  # jax-ndarray
Otherwise our pytype checker will complain.

Comment (shawnwang18) [reply praxis/optimizers.py]: Thanks, fixed.

**Comment by nouiz** [comment]
Why this PR isn't merged after approval?

**Comment by nouiz** [comment]
What is the perf impact that was seen internally?
In theory, there shouldn't be one.
Do you have any idea where it come from? How hard it would be to optimize it?

**Comment by nouiz** [comment]
@sherhut

**Comment by sherhut** [comment]
A way out here to unblock this would be to make the insertion of `cond` vs. `while` a params. Something like `use_cond_instead_of_while` and switch between the two. The default should be `False` for now.

**Comment by shawnwang18** [comment]
I have tried for a while to update the PR with a param to control whether to use `cond` or `while` operator, but I only understand the local pieces of the code, no idea on how to add the control param following praxis usual way. 

Could some guy from praxis team help make this update?  

@sherhut

**Comment by shawnwang18** [comment]
I don't which case in what case the using `while` to simulate `cond` operator could have better perf, but if we have that case, we should triage why `cond` operator have bad perf, it looks wired to use `while` to simulate the `cond` operator.

**Comment by bchetioui** [comment]
Not in the Praxis team, but @sherhut asked me to take a look. I would imagine that the right way to add the flag (at HEAD) would be to add a field in [ShardedStaticAccumulator](https://github.com/google/praxis/blob/59e990c2a916fe1be6002db2098fd7e2555a876d/praxis/optimizers.py#L2694) and in [sharded_static_accumulation](https://github.com/google/praxis/blob/59e990c2a916fe1be6002db2098fd7e2555a876d/praxis/optimizers.py#L2569C5-L2569C32) (where the branching should now happen). Then, just pass it to the call to `sharded_static_accumulation` within `ShardedStaticAccumulator`, and initialize `ShardedStaticAccumulator` as needed in whichever model you are using :)

File: praxis/optimizers.py
Comment (bchetioui) [inline praxis/optimizers.py:2826]: I think you'll also have to document the new class attribute.

File: praxis/optimizers.py
Comment (bchetioui) [inline praxis/optimizers.py:2695]: I think this should also probably be `false` by default.

File: praxis/optimizers.py
Comment (joker-eph) [inline praxis/optimizers.py:2767]: Is there an open bug with XLA? If not could one be filed on https://github.com/openxla/xla/issues ?

Comment (shawnwang18) [reply praxis/optimizers.py:2767]: I don't know the case, this is the original comments that describes the problem, but I don't see it in PAXML model.

**Comment by shawnwang18** [comment]
@akuegel  Can this PR be merged now?

**Comment by akuegel** [comment]
> @akuegel Can this PR be merged now?

I have no saying in this. This is in a code base where others need to approve.

@zhangqiaorjc Do the recent changes look ok now so that this PR can be merged?

**Comment by thomasjoerg** [comment]
Please resolve conflicts. I'll help with getting the PR merged once done.

**Comment by thomasjoerg** [comment]
Actually, I'll have to defer to @zhangqiaorjc . Looks like cannot shepperd this PR in. I didn't realize this was for google/praxis initially and I don't seem to have the required permissions here.

**Comment by shawnwang18** [comment]
@zhangqiaorjc could you check if the PR can be merged now?

**Comment by zhangqiaorjc** [comment]
adding the flag to gate this behavior LGTM

**Pull request merged**
